text/0000-upx-buildpack.md

```diff
@@ -0,0 +1,50 @@
+# UPX Support
+
+## Summary
+
+Introduce an UPX buildpack. This buildpack will provide the `upx` binary should any buildpacks want to use it. The `upx` binary can be used to compress executables by buildpacks producing binary executables and reduce binary sizes by often one half.
+
+The UPX buildpack would be maintained by the existing [Utilities team](https://github.com/orgs/paketo-buildpacks/teams/utilities). It uses libpak & libcnb.
+
+## Motivation
+
+There are a number of buildpacks where compressing generated executables can help to generate smaller images. For example, Java native image binaries when using the tiny builder can easily become larger than the entire rest of the image. Being able to compress a Java native image binary will help to reduce the image size in a significant way.
+
+This may be desirable with other buildpacks as well, such as the Go and Rust families, where good compression rates are also feasible.
+
+## Detailed Explanation
+
+1. A UPX buildpack has been developed. It will be contributed to Paketo.
+2. Other buildpacks wishing to compress binaries will opt-in by doing the following:
+   1. At detect time, require `upx`.
+   2. At build time after generating a binary, run `upx -q -9 <binary>` to compress the binary. This replaces the binary in-place with the compressed version.
+3. The native-image buildpack will implement this opt-in behavior.
+
+It is recommended that buildpacks which opt-in to using `upx`, make using `upx` optional with a default of disabled. This is because there is presently a known issue that causes a SEGFAULT when running a `upx` compressed binary on a Mac M1 Laptop running buildpacks under amd64 emulation. Defaulting to off provides for the largest compatibility.
```

Comment:
I think we should include the config option name here so that language families that choose to include this behavior can implement it consistently. Maybe `BP_ENABLE_COMPRESSION`?

Reply:
I added some text to try and address this. Let me know what you think.

1.) I went with `BP_COMPRESS_BINARIES` as I wanted to make sure it's clear what is being compressed. I'm open to other ideas, like if we want to keep it more passive like `BP_BINARY_COMPRESSION_METHOD`.

2.) I wrote it to contain the type of encryption to use, rather than be a true/false boolean flag. That way other types can be supported without requiring another env variable to specify the type.

Reply:
I'd prefer `BP_BINARY_COMPRESSION_METHOD` with a default of `none`. I think that variable name is the most intention revealing.

Reply:
Sounds good. I made that change. Also, I clarified that unset, empty or a value of `none` indicate no compression should be used.